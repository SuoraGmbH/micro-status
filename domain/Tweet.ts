export interface Tweet {
  userId: string;
  message: string;
  date: string;
  id: number;
}
