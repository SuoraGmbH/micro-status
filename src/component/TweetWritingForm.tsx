import * as React from 'react';
import { ChangeEvent, FormEvent } from 'react';
import { TweetWithoutId } from '../domain/Tweet';

interface Props {
  userId: string;
  addTweet: (tweet: TweetWithoutId) => any;
}

interface State {
  tweetText: string;
}

export class TweetWritingForm extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { tweetText: '' };
  }

  public render() {
    const { tweetText } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          type="text"
          name="text"
          value={tweetText}
          onChange={this.handleInputChange}
        />
        <button type="submit">
          <span>Tweet</span>
        </button>
      </form>
    );
  }

  private handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    if (this.state.tweetText.trim() === '') {
      event.preventDefault();
      return;
    }

    this.props.addTweet({
      date: new Date().toISOString(),
      userId: this.props.userId,
      message: this.state.tweetText,
    });

    this.setState({ tweetText: '' });
    event.preventDefault();
  };

  private handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      tweetText: event.target.value,
    });
  };
}
