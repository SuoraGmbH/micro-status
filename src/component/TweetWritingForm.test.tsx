import { mount } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { TweetWithoutId } from '../../domain/Tweet';
import { TweetWritingForm } from './TweetWritingForm';

it('renders correctly', () => {
  const addTweet = () => {};
  const tree = renderer
    .create(<TweetWritingForm userId="muhdiekuh" addTweet={addTweet} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('typing a value and submitting the form will result in a new tweet', () => {
  const addTweetCallback = jest.fn();

  const wrapper = mount(
    <TweetWritingForm userId="muhdiekuh" addTweet={addTweetCallback} />,
  );
  wrapper
    .find('input[name="text"]')
    .simulate('change', { target: { value: 'Hello World!' } });

  // Ensure value is in the field
  expect(wrapper).toMatchSnapshot();

  wrapper.find('input[type="submit"]').simulate('submit');

  expect(addTweetCallback.mock.calls.length).toBe(1);
  const tweet: TweetWithoutId = addTweetCallback.mock.calls[0][0];
  expect(tweet.userId).toBe('muhdiekuh');
  expect(tweet.message).toBe('Hello World!');
});
