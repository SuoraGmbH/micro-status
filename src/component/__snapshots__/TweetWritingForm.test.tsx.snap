// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="text"
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    type="submit"
  />
</form>
`;

exports[`typing a value and submitting the form will result in a new tweet: field filled 1`] = `
<TweetWritingForm
  addTweet={[MockFunction]}
  userId="muhdiekuh"
>
  <form
    onSubmit={[Function]}
  >
    <input
      name="text"
      onChange={[Function]}
      type="text"
      value="Hello World!"
    />
    <input
      type="submit"
    />
  </form>
</TweetWritingForm>
`;
